### DISCUSSION TITLE
`R`code chunk using `knitr::asis_output` and `output: asis` producing enclosing container (`.cell-output-display`)

### DISCUSSION THREAD

**Discussion by r-cheologist** [quarto-dev/quarto-cli#6874] [Q&A] [answered]
### Description

Consider a `_quarto.yml` like this:

```{qmd}
project: 
  type: default

execute: 
  engine: knitr
  echo: false
  output: asis

format:
  pdf:
    keep-md: true
```
and this `index.qmd`:

````{r}
---
engine: knitr # Required to execute inline code in the absence of code chunks
---

First Name:

: **Some**

Family Name:

: **One**

```{r}
requireNamespace("knitr")
knitr::asis_output(paste0(
    "Date of Birth:\n",
    ": Date"))
```

Email:

: [some.one@something.md](mailto:some.one@something.md)
````

Rendering this using `quarto render --to pdf` (on the surface) produces the expected results, but examining the generated `*.md` makes it obvious, that the chunk-produced output produces an enclosing container (`.cell-output-display`).

The detrimental effect of that become apparent if one adds the following to `_quarto.yml`:

```{qmd}
format:
  pdf:
    include-in-header: custom_tex.tex
```
which references this file:

```{latex}
\usepackage{enumitem}
\setlist[description]{nosep,style=multiline,leftmargin=3.5cm,font=\normalfont\textsf}
```
Rendering as above now produces tight `description` spacing for the first two description items (which are **not** separated by the enclosing container) but not the other two.

Can anyone hint at whether this is a bug, given that `output: asis` is [documented](https://quarto.org/docs/reference/cells/cells-knitr.html) as `Treat output as raw markdown with no enclosing containers.`?

How might I overcome this?

**Comment by mcanouil** [comment]
What's that? It's invalid markup ` ```{} # REMOVE "{}" - required for chunk representation in github md `.

Could you simplify your example? is the `_quarto.yml` really necessary?

Also you are mixing several things such as `knitr::asis_output()` and `output: asis`.
The former is a knitr function which is not really intended to be used with `output: asis` in knitr.
The second means text output has to be treated as is.

You can share a Quarto document using the following syntax, _i.e._, using more backticks than you have in your document (usually four ` ```` `).

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
---

This is a reproducible Quarto document using `format: html`.
It is written in Markdown and contains embedded R code.
When you run the code, it will produce a plot.

```{r}
plot(cars)
```

The end.
````
`````

**Reply by r-cheologist** [reply]
Thank you for that hint. Edited.

**Reply by mcanouil** [reply, marked as answer]
Could you try the following? (I don't have access to a proper LaTeX installation right now)

````qmd
---
format: pdf
include-in-header:
  - text: |
      \usepackage{enumitem}
      \setlist[description]{nosep,style=multiline,leftmargin=3.5cm,font=\normalfont\textsf}
keep-md: true
execute:
  echo: false
  output: asis
---

First Name:

: **Some**

Family Name:

: **One**

```{r}
cat(paste0("Date of Birth:\n", ": Date"))
```

Email:

: [some.one@something.md](mailto:some.one@something.md)
````

**Reply by r-cheologist** [reply]
That is indeed working as advertised. Now I have to revisit the [need](https://github.com/quarto-dev/quarto-cli/discussions/6542#discussion-5525921) to use `knitr::asis_output` in the first place...

**Comment by cderv** [comment]
I am wondering if we are correctly doing the right thing in Quarto for `output_asis`... it should be equivalent to `output: asis`  like in R Markdown context. 

````markdown
---
output: 
  pdf_document:
    keep_md: true
header-includes: |
      \usepackage{enumitem}
      \setlist[description]{nosep,style=multiline,leftmargin=3.5cm,font=\normalfont\textsf}
---

```{r, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```


First Name:

: **Some**

Family Name:

: **One**

```{r, results='asis'}
cat(paste0("Date of Birth:\n", ": Date"))
```

Email:

: [some.one@something.md](mailto:some.one@something.md)


First Name:

: **Some**

Family Name:

: **One**

```{r}
requireNamespace("knitr")
knitr::asis_output(paste0(
    "Date of Birth:\n",
    ": Date"))
```

Email:

: [some.one@something.md](mailto:some.one@something.md)

````

There shouldn't be any difference 🤔 

Also regarding this example, why not prefer inline chunk in this  case ? 

````

First Name:

: **Some**

Family Name:

: **One**

`r paste0("Date of Birth:\n", ": Date")`

Email:

: [some.one@something.md](mailto:some.one@something.md)
````

**Reply by cderv** [reply]
> There shouldn't be any difference

It seems in Quarto we have a difference and `knitr::asis_output` is not equivalent to `cat()` with `#| output: asis` 

We'll try to remember why is that - this will require documentation

**Reply by allefeld** [reply]
I have been struggling with that. It has other side effects, e.g. headings in generated Markdown are not picked up for the TOC because they are encapsulated in the div. It would be great if there was a way from within R code to achieve the equivalent of `#| output: asis`.

**Reply by cderv** [reply]
This is tracked here: 
- https://github.com/quarto-dev/quarto-cli/issues/9185

It is supposed to be fixed in **knitr** but somehow it seems to require some adaptation our quarto side. I'll handle this for 1.7

**Comment by r-cheologist** [comment]
On September 21, 2023 1:56:01 PM UTC, Christophe Dervieux ***@***.***> wrote:
>I am wondering if we are correctly doing the right thing in Quarto for `output_asis`... it should be equivalent to `output: asis`  like in R Markdown context. 
>
>````markdown
>---
>output: 
>  pdf_document:
>    keep_md: true
>header-includes: |
>      \usepackage{enumitem}
>      \setlist[description]{nosep,style=multiline,leftmargin=3.5cm,font=\normalfont\textsf}
>---
>
>```{r, include=FALSE}
>knitr::opts_chunk$set(echo = FALSE)
>```
>
>
>First Name:
>
>: **Some**
>
>Family Name:
>
>: **One**
>
>```{r, results='asis'}
>cat(paste0("Date of Birth:\n", ": Date"))
>```
>
>Email:
>
>: ***@***.******@***.***)
>
>
>First Name:
>
>: **Some**
>
>Family Name:
>
>: **One**
>
>```{r}
>requireNamespace("knitr")
>knitr::asis_output(paste0(
>    "Date of Birth:\n",
>    ": Date"))
>```
>
>Email:
>
>: ***@***.******@***.***)
>
>````markdown
>
>There should be any difference 🤔 
>
>Also regarding this example, why not prefer inline chunk in this  case ? 
>
>````
>
>First Name:
>
>: **Some**
>
>Family Name:
>
>: **One**
>
>`r paste0("Date of Birth:\n", ": Date")`
>
>Email:
>
>: ***@***.******@***.***)
>````
Thanks for picking this up. You are right, of course, with the advise to Inline in this case, but it is just a reproducible example to a more complex real world application, where conditional chunk inclusion comes in really Handy.

**Reply by cderv** [reply]
> but it is just a reproducible example to a more complex real world application, where conditional chunk inclusion comes in really Handy.

I understand. I would love to see this real example ! 

Other comment when generating markdown, especially using chunks (as they are closer to block content). When needing to mix raw markdown and computed content using a chunk (when inline not possible), it is sometimes best to use `knit_child` using external document, or template string. 

* https://bookdown.org/yihui/rmarkdown-cookbook/knit-expand.html
* https://bookdown.org/yihui/rmarkdown-cookbook/child-document.html

This allows to write Markdown context mixing R code, possibly inline, but write the inclusion of this in a chunk so that some options can be applied (like conditional insertion). 

As Markdown can be really sensible to syntax (like new line or other content in list), insuring that the whole block comes together is best. Because using a chunk (which is block element really), in the middle of another block markdown content like List can be tricky. 

Anyhow, several solutions as always, just sharing experience.

**Comment by r-cheologist** [comment]
On September 21, 2023 2:47:19 PM UTC, Christophe Dervieux ***@***.***> wrote:
>> but it is just a reproducible example to a more complex real world application, where conditional chunk inclusion comes in really Handy.
>
>I understand. I would love to see this real example ! 
It's actually my CV, which uses multiple interlocking profiles that govern chunk inclusion: 2 alternative languages, a long version and a one page resume, elements that cater to peculiarities of different application environments (private address yes or no, inclusion of trusted colleagues as reference yes or no, inclusion of earlier successful selection process yes or no) etc. 


>Other comment when generating markdown, especially using chunks (as they are closer to block content). When needing to mix raw markdown and computed content using a chunk (when inline not possible), it is sometimes best to use `knit_child` using external document, or template string. 
>
>* https://bookdown.org/yihui/rmarkdown-cookbook/knit-expand.html
>* https://bookdown.org/yihui/rmarkdown-cookbook/child-document.html
>
>This allows to write Markdown context mixing R code, possibly inline, but write the inclusion of this in a chunk so that some options can be applied (like conditional insertion). 
>
>As Markdown can be really sensible to syntax (like new line or other content in list), insuring that the whole block comes together is best. Because using a chunk (which is block element really), in the middle of another block markdown content like List can be tricky. 
>
>Anyhow, several solutions as always, just sharing experience. 
Interesting. I shall explore.
